src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +330,169 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
             { WasCompilerGenerated = true };
         }
 
+        // A synthesized entrypoint that forwards all calls to an async Main Method
+        internal sealed class AsyncForwardEntryPoint : SynthesizedEntryPointSymbol
+        {
+            // The user-defined asyncrhonous main method.
+            private MethodSymbol _userMain;
+            private DiagnosticBag _diagnosticBag;
+            private CSharpCompilation _compilation;
+
+            private ReturnKind _returnKind;
+            private ParamKind _paramKind;
+
+            private ImmutableArray<ParameterSymbol> _parameters;
+
+            // Which "kind" of Main is this: `void Main` or `int Main`?
+            // This is determined by looking at the return type of the user-provided Main
+            private enum ReturnKind
+            {
+                VoidReturning,
+                IntReturning,
+            }
+            // Does this main method need to forward arguments to the user-provided Main?
+            private enum ParamKind
+            {
+                NoArgs,
+                StringArrayArgs,
+            }
+
+            // Task -> Void
+            // Task<_> -> Int32
+            private static TypeSymbol TranslateReturnType(CSharpCompilation compilation, TypeSymbol returnType)
+            {
+                if (returnType.IsGenericTaskType(compilation)) {
+
+                    return compilation.GetSpecialType(SpecialType.System_Int32);
+                }
+                else
+                {
+                    return compilation.GetSpecialType(SpecialType.System_Void);
+                }
```

Comment:
What's the style formatting for breaking this across multiple lines?

I'm partial to

```
                return returnType.IsGenericTaskType(compilation) 
                    ? compilation.GetSpecialType(SpecialType.System_Int32)
                    : compilation.GetSpecialType(SpecialType.System_Void);
```